### PR TITLE
Update AutoProjectedGradientDescent

### PR DESCRIPTION
**Pull request by beat-buesser**
Signed-off-by: Beat Buesser <beat.buesser@ie.ibm.com>

# Description

This pull request provides updates for `AutoProjectedGradientDescent`. It extends `AutoProjectedGradientDescent` to run with estimators of all frameworks using the estimator's loss function requiring the estimator to predict logits and provide loss gradients. It updates `AutoProjectedGradientDescent` to use the new method `loss` of the Estimator API replacing custom functions in `AutoProjectedGradientDescent` and implements method `loss` for `TensorFlowClassifer`. It disables the loss type `difference_logits_ratio` for `TensorFlowClassifer`, for now until a suspected bug in that loss function has been fixed, by commenting the implementation as it seems to fail to select the second largest logit for cases where the largest logit is the true logit which occurs in small number of predictions, this only affects `AutoProjectedGradientDescent` with `TensorFlowClassifer`.)

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/estimators/classification/keras.py [hunk 1]
```diff
@@ -300,14 +300,18 @@ def _initialize_params(
         # Get the internal layer
         self._layer_names = self._get_layers()
 
-    def loss(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
+    def loss(self, x: np.ndarray, y: np.ndarray, reduction: str = "none", **kwargs) -> np.ndarray:
```
Comment (hesseltuinhof) [inline art/estimators/classification/keras.py:303]: Have you also considered using `reduction: Optional[str] = None` here?

File: art/attacks/evasion/auto_projected_gradient_descent.py [hunk 2]
```diff
@@ -89,258 +89,255 @@ def __init__(
...
+                        #     z_i = np.diag(z_i)
+                        #     z_y = np.diag(z_y)
+                        #
+                        #     dlr = -(z_y - z_i) / (z_1 - z_3)
+                        #
+                        #     return np.mean(dlr)
+                        #
+                        # self._loss_fn = loss_fn
+                        # self._loss_object = difference_logits_ratio(y_true=estimator._labels_ph,
+                        #                                             y_pred=estimator._output)
```
Comment (hesseltuinhof) [inline art/attacks/evasion/auto_projected_gradient_descent.py:182]: Looks like this code is replaced by the two `difference_logits_ratio` classes. Can we remove these comments here?

**Review by hesseltuinhof — CHANGES_REQUESTED** [review]
I have added two comments. Otherwise, the PR looks ready for merging.

File: art/estimators/classification/keras.py (hunk 1 above)
Comment (beat-buesser) [reply art/estimators/classification/keras.py:303]: Yes, I considered it. The reason for `"none"` and the options for `reductions was that it is what PyTorch is using for its loss functions.

File: art/attacks/evasion/auto_projected_gradient_descent.py (hunk 2 above)
Comment (beat-buesser) [reply art/attacks/evasion/auto_projected_gradient_descent.py:182]: Not completely, the other two `difference_logits_ratio` classes are specific for PyTorch and TensorFlow v2. The commented code here would be the class specific for TensorFlow v1, but it seems to contain a small bug as described in the exception above it and we keep the code commented until it is fix.

**Comment by beat-buesser** [comment]
Hi @hesseltuinhof Thank you very much for your review, I have responded to your comments above.

File: art/estimators/classification/keras.py (hunk 1 above)
Comment (hesseltuinhof) [reply art/estimators/classification/keras.py:303]: That's a good argument 👍

**Review by hesseltuinhof — APPROVED** [review]
